R/lm_lin.R

```diff
@@ -61,38 +61,56 @@ lm_lin <- function(formula,
   ))
 
   outcome <- model_data$outcome
+  n <- length(outcome)
   design_matrix <- model_data$design_matrix
   weights <- model_data$weights
   cluster <- model_data$cluster
 
-  # If Z is a factor, can't use variable name
-  # So get first non-intercept column (always will be treatment)
+  # Get treatment columns
+  has_intercept <- attr(terms(formula), "intercept")
   treat_col <- which(attr(design_matrix, "assign") == 1)
-  treat_name <- colnames(design_matrix)[treat_col]
-
-  # allow drop to 1 dim if well specified
-  # this allows * operator to "sweep" across demeaned covars later
-  treatment <- design_matrix[, treat_col]
-
-  if (is.matrix(treatment)) {
-    stop(
-      "Treatment variable must have no more than two levels in the data (or the subset you are using)."
-    )
-  }
+  treatment <- design_matrix[, treat_col, drop = FALSE]
+  design_mat_treatment <- colnames(design_matrix)[treat_col]
 
+  # Check case where treatment is not factor and is not binary
   if (any(!(treatment %in% c(0, 1)))) {
-    stop(
-      "Treatment variable must be binary for the Lin estimator."
-    )
+    if (ncol(treatment) > 1) {
+      stop(
+        "Treatment variable must be binary or a factor for the Lin estimator."
+      )
+    } else {
+      # create dummies for non-factor treatment variable
+
+      # Drop out first group if there is an intercept
+      vals <- sort(unique(treatment))[if(has_intercept) {-1} else {TRUE}]
+      n_treats <- length(vals)
+      # TODO warn if too many values?
+
+      treatment_mat <- matrix(
+        NA,
+        nrow = n,
+        ncol = n_treats,
+        dimnames = list(NULL,
+                        paste0(colnames(design_matrix)[treat_col], vals))
+      )
+
+      for (i in 1:n_treats) {
+        treatment_mat[, i] <- as.numeric(treatment == vals[i])
```

Comment:
Can vectorize the loop with outer - ```treatment_mat[] <- outer(treatment, vals, `==`) ```

Reply:
Note that treatment is a matrix with multiple columns!

Reply:
wait, if `treatmat_mat[, i]` is not the same length as `as.numeric(treatment == vals[i])` that also seems bad.

Reply:
let's sleep on this and make it less idiosyncratic.